### ISSUE TITLE
Bug with dispatching events to invoked or spawned actors

### ISSUE THREAD

**Issue by akfernun** [LevelbossMike/ember-statecharts#372]
## Description

Today, Xstate provides functionality for spawning or invoking actors within a machine. When doing so with ember-statecharts, subsequent events that are sent to the spawned/invoked actor do not trigger re-renders. So for example, if I have a machine that invokes an actor, and I try to send new events to that actor, ember will not render changes caused by the events. 

## Example

```js 
// Invoked child machine
const minuteMachine = createMachine({
  id: 'timer',
  initial: 'active',
  states: {
    active: {
      after: {
        60000: { target: 'finished' }
      }
    },
    finished: { type: 'final' }
  }
});

const parentMachine = createMachine({
  id: 'parent',
  initial: 'pending',
  states: {
    pending: {
      invoke: {
        src: minuteMachine, // Unfortunately, if we send events to this machine once it is invoked, Ember will 
        // not re-render because the interpreter that is created is an Xstate interpreter, not an ember-statecharts
       // interpreter
        onDone: 'timesUp'
      }
    },
    timesUp: {
      type: 'final'
    }
  }
});
```

#### ember-statecharts

https://ember-twiddle.com/b69c24fb22daf2cf0ebd71cabe08b425?openFiles=components.parent-machine%5C.js%2C

#### react
See this example for how things would ideally work. 

https://codesandbox.io/s/child-state-machines-t8cy2?file=/src/machine.js

## Proposal

This functionality works today in the React and Vue Xstate libraries through the use of the `useActor` hook/api. Would it be possible to implement a similar API for ember-statecharts such that spawned or invoked actors will trigger DOM re-renders.

## Related resources

https://xstate.js.org/docs/packages/xstate-react/#useactor-actor-getsnapshot
https://xstate.js.org/docs/packages/xstate-vue/#useactor-actor-getsnapshot
https://xstate.js.org/docs/guides/actors.html#actors
https://xstate.js.org/docs/guides/communication.html#invoking-machines

**Comment by NullVoxPopuli** [comment]
I have a working solution for this after pairing with and learning from @akfernun a bit yesterday about this pattern with xstate.
It wasn't too much code, @LevelbossMike, and idk if you want to straight copy it, or maybe i should make a micro lib for 'tracked-xtate-interpreters'?

https://github.com/NullVoxPopuli/ember-statechart-component/blob/main/ember-statechart-component/src/-private/proxy.ts#L52

Tldr: when an actor is created, and that actor is an interpreter, it gets transparently wrapped in reactivity

**Comment by LevelbossMike** [comment]
@NullVoxPopuli @akfernun Thanks for bringing this up 👍. Great that you already have a solution in place. I'll look into the solution and try to recreate it including tests. Not sure when I'll be able to tackle this though - anybody of you two interested in submitting this including tests?

**Comment by LevelbossMike** [comment]
@akfernun @NullVoxPopuli After looking into the xstate docs for `invoke` and `actor` I have changed my mind on this and don't think this is something that we need to expand the api surface of ember-statecharts to support this use-case. XState already has you covered:

According to the [invoke](https://xstate.js.org/docs/guides/communication.html#invoking-services)-docs 

> For machines to communicate with each other, the parent machine invokes a child machine and __listens to events sent from the child machine__ via sendParent(...), __or waits for the child machine to reach its final state__, which will then cause the onDone transition to be taken.

In my mind you can handle the challenge you are describing where you invoke a machine by sending an event to the parent which will trigger the `onTransition`-hook and update the parent-machine's state. I totally get the idea that you want to be able to access the state of the invoked machine but this can be handled already - it doesn't look like we need to extend the api interface to do so.

For spawning actors [something similar applies](https://xstate.js.org/docs/guides/actors.html#syncing-and-reading-state):

> One of the main tenets of the Actor model is that actor state is private and local - it is never shared unless the actor chooses to share it, via message passing. Sticking with this model, an actor can notify its parent whenever its state changes by sending it a special "update" event with its latest state. In other words, parent actors can subscribe to their child actors' states.

But again if need be you can opt into keeping up to date - i.e. when you are really interested in the states of each individual actor and you want the statechart to reactively keep in sync you can pass the `{ sync: true }`-option when spawning an actor which will give you the desired behavior and keep the parent in sync with what is happening on its children or you can tell the actor explicitly to send an update via [sendUpdate](https://xstate.js.org/docs/guides/actors.html#sending-updates) to trigger an update.

Here's an [ember-twiddle](https://ember-twiddle.com/1bc630348f9890e68de73d551e396238?openFiles=components.parent-machine%5C.js%2C) that demonstrates how I suggest to handle a situation like you describe. A parent opts into the sync and will then be able to access the always up to date state in a getter.

I left out the invoke example because my suggestion there would be to send an event back to the parent which will also trigger the state update like any other event.

I hope this solves your issue. I'm closing this as this is already something that ember-statecharts support we just have to adapt our machines to give us the expected behavior. I'm also adding the documentation label because this is certainly something that would be beneficial to document. Are you interested in trying the suggested solution on your challenge and to submit an update to the docs page for this @akfernun ?

**Comment by NullVoxPopuli** [comment]
There is no api surface change with the solution we came up with though 🤷

**Comment by LevelbossMike** [comment]
@NullVoxPopuli that's right - no api surface change but an internal change that proxies an XState internal interpreter property `children`. What you were trying to do is already possible with the way ember-statecharts works today, you can get the expected behavior by relying on XState features e.g. `{ sync: true }` - what do you want to see added and why?

**Comment by akfernun** [comment]
@LevelbossMike, I might be able to shed a little bit of light here. While yes, it is technically possible to accomplish this particular thing with `{ sync: true }`, this isn't considered to be a great practice. The following excerpt is taken from the same Xstate docs: 

> Prefer sending events to the parent explicitly (sendUpdate()) rather than subscribing to every state change. Syncing with spawned machines can result in "chatty" event logs, since every update from the child results in a new "xstate.update" event sent from the child to the parent.

As a follow-up to this, I don't think it's good practice to start introducing things into state machines with the _sole_ reason being to achieve reactivity. In general, state machines themselves should not be aware of the context in which they're operating, whether that be within a JS framework like Ember or just a plain ole JS file. In this case, using `{ sync: true }` feels like more of a work-around than a solution. 

One thing to note here is that, this concept is not something that's original or something that's just a "nice-to-have". Rather, this functionality is something [that is supported by the `xstate-react`](https://xstate.js.org/docs/packages/xstate-react/#useactor-actor-getsnapshot) and `xstate-vue` libraries themselves (which are maintained by core authors of xstate).

**Comment by LevelbossMike** [comment]
You can also `sendUpdate` in the example. `sync: true` is just an easier way for demoing this. The important part being that the object that uses `useMachine` updates its state.

**Comment by LevelbossMike** [comment]
That being said I'd be happy to merge a PR for `useActor` that allows passing a ref to an actor that adds reactivity to it. Personally I don't see the absolute need for it given the above reasoning but given the fact both of you seem to feel strongly about this and both of you think that's a good addition I'd be stoked to see a PR for this. @NullVoxPopuli @akfernun

**Comment by akfernun** [comment]
> I see why react would introduce useActor but Ember's reactivity model seems to work well for this use case already?!

Can you explain this line a little more please? I guess I'm confused on why React would need it but Ember wouldn't? If Ember's reactivity model worked well for this use case, then we wouldn't need to submit a PR for it

**Comment by LevelbossMike** [comment]
There is no real reactivity like we use in Ember with React, right?

I.e. they want to have a hook to access the state of the actor because that makes it easier for them to access the actors state.

```
get actorRef() {
  return this.statechart.state.context.actorRef;
}
```

Works when you `sync: True` or when you `sendUpdate`. To me that feels fine.

When I understand correctly you might want to use this like this without the need to `sync` or `sendUpdate`

```
get actor() {
  return useActor(this.statechart.state.context.actorRef);
}
```

We might be able to achieve this with the existing usable. And by adapting it a bit to add an onTransition hook to
An interpreter that xstate
Starts for you. Do you feel this is something you might want try to contribute? Happy to help with the implementation.

**Comment by akfernun** [comment]
@LevelbossMike, happy to try and implement something for this. It might be a little bit however. Time is a little short for me right now 😄
